File: js/packages/phoenix-client/src/utils/semverUtils.ts
/**
 * Lightweight semantic version utilities.
 */

import type { SemanticVersion } from "../types/semver";

// ---------------------------------------------------------------------------
// Parsing
// ---------------------------------------------------------------------------

/**
 * Parse a semver-like version string (e.g. "13.14.0") into a triple.
 * Returns `null` if the string is not a valid version.
 */
export function parseSemanticVersion(raw: string): SemanticVersion | null {
  const parts = raw.trim().split(".");
  if (parts.length < 3) return null;
  const major = parseInt(parts[0]!, 10);
  const minor = parseInt(parts[1]!, 10);
  const patch = parseInt(parts[2]!, 10);
  if (
    Number.isNaN(major) ||
    Number.isNaN(minor) ||
    Number.isNaN(patch) ||
    major < 0 ||
    minor < 0 ||
    patch < 0
  ) {
    return null;
  }
  return [major, minor, patch];
}

// ---------------------------------------------------------------------------
// Formatting
// ---------------------------------------------------------------------------

/**
 * Format a `SemanticVersion` as a `"major.minor.patch"` string.
 */
export function formatVersion(version: SemanticVersion): string {
  return `${version[0]}.${version[1]}.${version[2]}`;
}

// ---------------------------------------------------------------------------
// Comparison
// ---------------------------------------------------------------------------

/**
 * Returns `true` if `version` is greater than or equal to `minVersion`.
 */
export function satisfiesMinVersion({
  version,
  minVersion,
}: {
  version: SemanticVersion;
  minVersion: SemanticVersion;
}): boolean {
  const [serverVersionMajor, serverVersionMinor, serverVersionPatch] = version;
  const [minVersionMajor, minVersionMinor, minVersionPatch] = minVersion;
  if (serverVersionMajor !== minVersionMajor)
    return serverVersionMajor > minVersionMajor;
  if (serverVersionMinor !== minVersionMinor)
    return serverVersionMinor > minVersionMinor;
  return serverVersionPatch >= minVersionPatch;
}

File: packages/phoenix-client/src/phoenix/client/utils/semver_utils.py
"""Generic semantic version utilities.

Provides a ``SemanticVersion`` type alias, parsing, comparison, and
formatting helpers that are not tied to any specific versioned system.
"""

from __future__ import annotations

from typing import Optional, Tuple

SemanticVersion = Tuple[int, int, int]


def parse_semantic_version(raw: str) -> Optional[SemanticVersion]:
    """Parse a version string like ``'1.2.3'`` into a ``(1, 2, 3)`` tuple.

    Returns ``None`` if the string cannot be parsed.
    """
    try:
        parts = raw.strip().split(".")
        if len(parts) >= 3:
            return (int(parts[0]), int(parts[1]), int(parts[2]))
    except (ValueError, IndexError):
        pass
    return None


def satisfies_min_version(version: SemanticVersion, min_version: SemanticVersion) -> bool:
    """Return ``True`` if *version* is ``>=`` *min_version*."""
    return version >= min_version


def format_version(version: SemanticVersion) -> str:
    """Format a ``SemanticVersion`` tuple as a ``'major.minor.patch'`` string."""
    return f"{version[0]}.{version[1]}.{version[2]}"
